examples/drivers/windows/otLwf/settings.c

```diff
@@ -39,80 +39,497 @@ void otPlatSettingsInit(otInstance *otCtx)
+    *aValueLength = (uint16_t)pInfo->DataLength;
+    if (aValue)
+    {
+        memcpy(aValue, pInfo->Data, pInfo->DataLength);
+    }
+
+error:
+
+    if (pInfo) FILTER_FREE_MEM(pInfo);
+    if (regKey) ZwClose(regKey);
+
+    return status;
+}
+
+NTSTATUS FilterWriteSetting(_In_ PMS_FILTER pFilter, uint16_t aKey, int aIndex, const uint8_t *aValue, uint16_t aValueLength)
+{
+    HANDLE regKey = NULL;
+    OBJECT_ATTRIBUTES attributes;
+    DECLARE_UNICODE_STRING_SIZE(Name, 20);
+
+    // Convert 'aKey' to a string
+    RtlIntegerToUnicodeString((ULONG)aKey, 16, &Name);
+
+    InitializeObjectAttributes(
+        &attributes,
+        &Name,
+        OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
+        pFilter->otSettingsRegKey,
+        NULL);
+
+    // Create/Open the registry key
+    NTSTATUS status =
+        ZwCreateKey(
+            &regKey,
+            KEY_ALL_ACCESS,
+            &attributes,
+            0,
+            NULL,
+            REG_OPTION_NON_VOLATILE,
+            NULL);
+
+    NT_ASSERT(NT_SUCCESS(status));
+    if (!NT_SUCCESS(status))
+    {
+        LogError(DRIVER_DEFAULT, "ZwCreateKey for %S key failed, %!STATUS!", Name.Buffer, status);
+        goto error;
+    }
+
+    // Convert 'aIndex' to a string
+    RtlIntegerToUnicodeString((ULONG)aIndex, 16, &Name);
+
+    // Write the data to the registry
+    status =
+        ZwSetValueKey(
+            regKey,
+            &Name,
+            0,
+            REG_BINARY,
+            (PVOID)aValue,
+            aValueLength);
+
+    if (!NT_SUCCESS(status))
+    {
+        LogError(DRIVER_DEFAULT, "ZwSetValueKey for %S value failed, %!STATUS!", Name.Buffer, status);
+        goto error;
+    }
+
+error:
+
+    if (regKey) ZwClose(regKey);
+
+    return status;
+}
+
+NTSTATUS FilterDeleteSetting(_In_ PMS_FILTER pFilter, uint16_t aKey, int aIndex)
+{
+    HANDLE regKey = NULL;
+    OBJECT_ATTRIBUTES attributes;
+    DECLARE_UNICODE_STRING_SIZE(Name, 20);
+
+    // Convert 'aKey' to a string
+    RtlIntegerToUnicodeString((ULONG)aKey, 16, &Name);
+
+    InitializeObjectAttributes(
+        &attributes,
+        &Name,
+        OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
+        pFilter->otSettingsRegKey,
+        NULL);
+
+    // Open the registry key
+    NTSTATUS status =
+        ZwOpenKey(
+            &regKey,
+            KEY_ALL_ACCESS,
+            &attributes);
+
+    if (!NT_SUCCESS(status))
+    {
+        // Key doesn't exist
+        goto error;
+    }
+
+    // If 'aIndex' is -1 then delete the whole key, otherwise delete the individual value
+    if (aIndex == -1)
+    {
+        // Delete the registry key
+        status = ZwDeleteKey(regKey);
+    }
+    else
+    {
+        UCHAR KeyInfoBuffer[128] = { 0 };
+        PKEY_FULL_INFORMATION pKeyInfo = (PKEY_FULL_INFORMATION)KeyInfoBuffer;
+        ULONG KeyInfoLength = sizeof(KeyInfoBuffer);
+
+        // When deleting an individual value, since order doesn't matter, we will actually
```

Comment:
Well, the code I have here works for that example. :). But with all seriousness, the only code that calls `otPlatSettingsDelete` for an element in an array is `MleRouter::RemoveStoredChild`. It just iterated through the settings until it finds the one it wants and deletes it.

@jwhui Do we really need to preserve order in the delete? If so, then why not in `otPlatSettingsAdd`? I would ask that we relax the requirements on delete, otherwise it could become very expensive if there are a lot of items after that one in the array.